### PR TITLE
Multi-GPU training, DDP

### PR DESCRIPTION
**Pull request by geronimi73**
Tried to keep it simple.

## Changes (+65 lines)
* Added DDP detection and handling to `train.py` directly, run with `torchrun --nproc-per-node num_gpus train.py` instead of `python train.py`. Again, I tried to keep it simple, and decided to **not** add another file `train_ddp.py`.
* Functions `init_dist`, `destroy_dist` etc. in `train.py` could be moved to a separate file, left them in `train.py` for now.
* Batch size `train_cfg.batch_size` used as **per device** batch size rather than dividing `train_cfg.batch_size` by the number of GPUs. In other words, effective batch size is `train_cfg.batch_size * num_gpus`. 
* Metrics `batch_loss`, `tokens_per_second`, `avg_train_loss`, `total_tokens_processed` are gathered from all ranks instead of guesstimating global from local. I checked with 1, 2, 4 GPUs and LGTM but, speaking from experience, I might have messed something up there, please check.
* Especially the above adds a lot of `if is_dist(): ..`s, tried to keep them single-line where possible but it still adds a few lines.
* Added `DistributedSampler`, always, even in non-DDP training run to save another `if .. else`

I think that's all of the changes, hope I didn't miss anything. 

## Speed test: 1x vs 2x vs 4x 3090 for a few seconds
* 2 epochs of split `ai2d` only
Screenshot 2025-05-10 at 09 43 20

## Accuracy test: 4x3090 
* bs 64/device, GPUs power cap 250w
* splits `"ai2d", "aokvqa", "chart2text", "chartqa", "clevr", "cocoqa", "datikz", "diagram_image_to_text", "docvqa", "dvqa", "figureqa", "finqa", "geomverse", "hateful_memes", "hitab", "iam", "iconqa", "infographic_vqa", "intergps", "localized_narratives", "mapqa", "multihiertt", "ocrvqa", "plotqa", "raven", "rendered_text", "robut_sqa", "robut_wikisql", "robut_wtq", "scienceqa", "screen2words", "st_vqa", "tabmwp", "tallyqa", "tat_qa", "textcaps", "textvqa", "tqa", "vistext", "visual7w", "visualmrc", "vqarad", "vqav2", "vsr", "websight"`

Screenshot 2025-05-11 at 19 39 28

## Related issues
https://github.com/huggingface/nanoVLM/issues/13

## and then
Thank you for this codebase ❤️❤️ I love minimal stuff!

### CODE REVIEW

**Comment by lusxvr** [comment]
Sweet stuff, thank you for the contribution! I will take a look at this :)

File: train.py [hunk 1]
```diff
@@ -3,9 +3,12 @@
 import wandb
 import argparse
 import torch.optim as optim
+from statistics import mean
```
Comment (andimarafioti) [inline train.py:6]: Super small nit, but why use statistics and not torch

Comment (geronimi73) [reply train.py:6]: `torch.mean` needs a tensor but we got a list

Comment (lusxvr) [reply train.py:6]: Do you syncronize the whole object in your `dist_gather`? From what I tried (without releasing) and saw online a common approach is to only reduce/sync the individual tensors when you need it like
`# Synchronize loss across all processes
                loss_tensor = torch.tensor(batch_loss).to(device)
                dist.all_reduce(loss_tensor, op=dist.ReduceOp.SUM)`

But not sure what the best way / best practice would be here

Comment (geronimi73) [reply train.py:6]: > Do you syncronize the whole object in your `dist_gather`?

yes. It's the same way 🤗 accelerate does it.
https://github.com/huggingface/accelerate/blob/d03664b8670d96c45238e904f54f477ab6df49fd/src/accelerate/utils/operations.py#L252

Comment (andimarafioti) [reply train.py:6]: Very cool, thanks for the explanation!

**Comment by lusxvr** [comment]
Did you experiment with `torch.compile()` in the DDP setup? When I run it without compiling it works fine and gives me a speedup of ~4x (1 H100 compiled vs 8 H100 uncompiled).

A quick google search brought up this: https://discuss.pytorch.org/t/combining-torch-compile-and-distributeddataparallel/184591 but we already compile before the wrapping.

Not a big issue, but would be interesting to understand.

**Comment by geronimi73** [comment]
> but when I try to compile the model it throws an expection in the subprocesses.

not till now. Just tried, and on torch `2.4.0+cu121` I see something different. I don't get an exception but a warning. But the training works fine.

The warning happens just when invoking `forward()` the *very first time* (presumably because that's [when the actual compilation happens](https://discuss.pytorch.org/t/compiling-the-model-results-in-20x-slow-down/185693/9)). Happens with (`torchrun .. train.py`) and without DDP (`python3 train.py`) tough.

Haven't found anything on *why* this warning happens unfortunately.

```
[rank0]:W0514 17:38:23.644000 128295424806912 torch/fx/experimental/symbolic_shapes.py:4449] [0/0_1] xindex is not in var_ranges, defaulting to unknown range.
[rank1]:W0514 17:38:26.377000 128643271733248 torch/fx/experimental/symbolic_shapes.py:4449] [0/0_1] xindex is not in var_ranges, defaulting to unknown range.
onto model done
[rank0]:W0514 17:38:55.397000 128289710540352 torch/fx/experimental/symbolic_shapes.py:4449] [5/0] xindex is not in var_ranges, defaulting to unknown range.
onto model done
[rank1]:W0514 17:39:06.998000 128637927949888 torch/fx/experimental/symbolic_shapes.py:4449] [5/0] xindex is not in var_ranges, defaulting to unknown range.
```

**Comment by lusxvr** [comment]
Hm okay, let me check what versions we are running on the cluster.

**Comment by geronimi73** [comment]
> there is a small conflict now

resolved and just started a test run. loss already looks totally different than my old 4x3090 run, I assume that's because of the new LR schedule and probably other PRs. This file/repo is blowing up quickly.

Will check back tomorrow and post the results of the test run.

**Comment by geronimi73** [comment]
LGTM. works technically, accuracy is worse though

Screenshot 2025-05-16 at 06 06 22

**Comment by lusxvr** [comment]
Cool, thanks for resolving! 

I will try it on our cluster. Sometimes there is quite some sensitivity between runs, this could also explain the dropped accuracy. The calculation of the performance on MMStar is a bit hacky at the moment.

We also noticed that the pure focus on the accuracy of one single downstream benchmark might not be that good of a signal (as @andimarafioti reminded me of Goodhart: “When a measure becomes a target, it ceases to be a good measure”). I therefore added the validation loss again, might be a more robust way to evaluate different settings.

File: train.py [hunk 2]
```diff
@@ -248,59 +293,69 @@ def train(train_cfg, vlm_cfg):
...
+    if is_master():
+        avg_epoch_time = sum(epoch_times) / len(epoch_times)
+        total_training_time = sum(epoch_times)
+        total_samples_processed = len(train_loader.dataset) * train_cfg.epochs
+        avg_time_per_sample = total_training_time / total_samples_processed
+        print(f"Average time per epoch: {avg_epoch_time:.2f}s")
+        print(f"Average time per sample: {avg_time_per_sample:.4f}s")
+
+        # unwrap the model for eval if DDP
+        accuracy = test_mmstar(model.module if is_dist() else model, tokenizer, test_loader, device)
```
Comment (lusxvr) [inline train.py:351]: This was not needed before the lr scheduling right? Do you know why it starts to hang now but did not before?

Comment (geronimi73) [reply train.py:351]: I had the `is_dist()` check inside `test_mmstar` but moved it outside because it looks more elegant to me. that's why we need to now check what we pass to `test_mmstar`

**Comment by lusxvr** [comment]
image


The results of my test are in, and we see some interesting things. 
Burgundy: single gpu, no lr schedule, no val set (best run so far, this is the uploaded checkpoint)
Green: single gpu, lr schedule, val set
Blue: multi gpu, no lr schedule, no val set
Red: multi gpu, lr schedule, val set

1. The accuracy is quite inconsistent. I believe this is because both our way to evaluate MMStar is not ideal, and due to the fact that it is a single downstream task for a very sparsely trained and very small VLM. Nonetheless, we should either improve the evaluation procedure or not take it as a strong signal since the variance seems to high.
2. LR scheduling seems to improve convergence of the accuracy compared to non LR scheduling
3. While the train loss in the DDP setup is a bit higher at the end compared to a single GPU (a bit unexpected), the val loss converges to the same value (good)
4. TpS suffers with the val set, since we unroll the distributed model and essentially make the whole training wait for one gpu to finish with the validation batches (this should be improved, as well as the distribution of the MMStar eval ideally)

Happy to discuss these plots and the take-aways you get from them!

**Comment by andimarafioti** [comment]
For 2, you should compare runs with the same batch size. Either do gradient accumulation when you only have one GPU, or use a smaller per_gpu batch size when you have several. For the comparison that should reduce a big factor for differences.
For 4, you can do the evaluation distributed. To start, I would just not evaluate MMStar as often.

**Comment by lusxvr** [comment]
@andimarafioti very true, thanks for catching that. When I matched the batch size for the different runs they perform basically the same. 

image

**Comment by lusxvr** [comment]
@geronimi73 do you mind giving me push access to your fork? I adapted the logic of the gradient accumulation steps as well as the logging intervals.

**Comment by geronimi73** [comment]
> * Which commit did you use?
> * Did you modify any configurations from the default settings?

I used my fork (=this PR): https://github.com/geronimi73/nanoVLM/tree/ddp 
config all default except:
```
batch_size: int = 64
compile: bool = False
train_dataset_name: tuple[str, ...] = ("ai2d", "aokvqa", "chart2text", "chartqa", "clevr", "cocoqa", "datikz", "diagram_image_to_text", "docvqa", "dvqa", "figureqa", "finqa", "geomverse", "hateful_memes", "hitab", "iam", "iconqa", "infographic_vqa", "intergps", "localized_narratives", "mapqa", "multihiertt", "ocrvqa", "plotqa", "raven", "rendered_text", "robut_sqa", "robut_wikisql", "robut_wtq", "scienceqa", "screen2words", "st_vqa", "tabmwp", "tallyqa", "tat_qa", "textcaps", "textvqa", "tqa", "vistext", "visual7w", "visualmrc", "vqarad", "vqav2", "vsr", "websight")
```

trained on 4x 3090
[wandb logs](https://wandb.ai/g-ronimo/nanoVLM/runs/btkul5wh)

**Comment by geronimi73** [comment]
> Happy to discuss these plots and the take-aways you get from them!

* I would definitely leave the accuracy in there. It's a good additional metric. 
* Agree with @andimarafioti, you can compare those metrics only for runs with the same global bs.
* Distributing the evals will add **a lot** of code. It was not my intent to rebuild 🤗 trainer + accelerate in a single file. I would not do that.

**Comment by lusxvr** [comment]
> Distributing the evals will add a lot of code. It was not my intent to rebuild 🤗 trainer + accelerate in a single file. I would not do that.

True, I left the MMStar to run on a single gpu with a bit of a reduced frequency, and parallelized the validation set. Since this is the same as the training it's quite manageable and still gives a nice speedup

**Comment by lusxvr** [comment]
I pushed some updated. Mainly there is now proper support for gradient accumulation, and I verified that the DDP setup produces the same results as the non DDP. Unfortunately for my torch version `torch.compile` still throws an error with DDP, but I will take care of this seperately.

Result 1:
image
When using the 'old' global batch size of 256, both DDP and single GPU runs now produce the same accuracy and loss, with the general accuracy at levels we expect.

Result 2:
image
When maximizing the individual GPU capacity (batch size 256) while training with DDP (meaning a global batch size of 8*256=2048, violet curve) we get the same results as for a single gpu with batch size 256 and 8 gradient accumulation steps (resulting in the same effective batch size of 2048, green curve). The accuracy for this is quite bad, but this is probably due to the high sensitivity of our MMStar Pipeline, and the higher global batch size without an adapted learning rate. I will run a quick ablation for this new batch size to figure out a better value.

Result 3:
I parallelized the validation similarly as the training and decreased the evaluation frequency of MMSta. With the same DDP setup as before, this gives quite significant total speedups again (red curve).
We are now at a speedup of ~6x, comparing the Single GPU with `torch.compile` and 8 GPUs without `torch.compile` (will investigate this issue for me separately since it works for you)

Result 4:
For some reason, using `torchrun` (instead of calling `train.py` ourselves) gives us way better reproducibility than we had without it (see loss curves in the second image). Since I was a bit worried about this in general, I will take this unexpected bonus :)

Please feel free to validate my changes again on your end, if there is no objection I will merge this!

**Comment by lusxvr** [comment]
@Nagi-ovo feel free to also try the new changes, for me the gradient accumulation when resulting in a total batch size of 256 produces the same results now.

**Comment by geronimi73** [comment]
> Please feel free to validate my changes again on your end, if there is no objection I will merge this!

I've noticed a few things, left comments. I think it works technically but this thing is getting really big and messy. There's a lot of potential for refactoring and it would probably be good to start adding tests.

File: train.py [hunk 3]
```diff
@@ -259,8 +260,20 @@ def train(train_cfg, vlm_cfg):
             labels = batch["labels"].to(device)
             attention_mask = batch["attention_mask"].to(device)
 
+            # if DDP: don't sync gradients when accumulating
```
Comment (andimarafioti) [inline train.py:263]: It isn't obvious why you do this, I would add this comment: 

```python
# When using DDP with gradient accumulation,
# skip gradient synchronization on intermediate steps to save time.
# Gradients only need to be synced at the end of each accumulation cycle.
```

**Review by andimarafioti — APPROVED** [review]
Left a small comment on something that isn't obvious to me.

Otherwise, I think this looks really good. Yes, it adds complexity, and there is room to improving that complexity, but I wouldn't keep dragging this PR. I think it's great as is and we can iterate on it.

**Comment by lusxvr** [comment]
Sweet, thank you for the quick adoption @geronimi73 !

**Review by lusxvr — APPROVED** [review]

**Comment by lusxvr** [comment]
Really appreciate your contributions here @geronimi73 !

We saw you even posted a blog about DDP recently, would love to shout you / it out on Twitter if you have it?

**Comment by StrongerXi** [comment]
Hi @lusxvr would you mind sharing the error you got with `torch.compile`?

**Comment by lusxvr** [comment]
Hi @StrongerXi, I created a separate issue for this (https://github.com/huggingface/nanoVLM/issues/61), I have all the details there
